app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/PeerGroupHandshakeAsyncTask.kt

```diff
@@ -0,0 +1,151 @@
+            return@withContext if (isKiwixHandshake(kiwixHandShakeMessage) && this.isActive) {
+              if (BuildConfig.DEBUG) {
+                Log.d(TAG, "Client IP address: " + server.inetAddress)
+              }
+              exchangeFileTransferMetadata(server.getOutputStream(), server.getInputStream())
+              server.inetAddress
+            } else {
+              // Selected device is not accepting wifi direct connections through the kiwix app
+              null
+            }
+          }
+      } catch (ex: Exception) {
+        ex.printStackTrace()
+        return@withContext null
+      }
+    } else if (wifiDirectManager.isGroupFormed && this.isActive) { // && !groupInfo.isGroupOwner
+      try {
+        Socket().use { client ->
+          client.reuseAddress = true
+          client.connect(
+            InetSocketAddress(
+              wifiDirectManager.groupOwnerAddress.hostAddress,
+              PEER_HANDSHAKE_PORT
+            ), 15000
+          )
+          val objectOutputStream = ObjectOutputStream(client.getOutputStream())
+          // Send message for the peer device to verify
+          objectOutputStream.writeObject(HANDSHAKE_MESSAGE)
+          exchangeFileTransferMetadata(client.getOutputStream(), client.getInputStream())
+          return@withContext wifiDirectManager.groupOwnerAddress
+        }
+      } catch (ex: Exception) {
+        ex.printStackTrace()
+        return@withContext null
+      }
+    }
+    return@withContext null
+  }
+
+  private fun isKiwixHandshake(handshakeMessage: Any): Boolean =
+    HANDSHAKE_MESSAGE == handshakeMessage
+
+  private fun exchangeFileTransferMetadata(outputStream: OutputStream, inputStream: InputStream) {
+    if (wifiDirectManager.isFileSender) {
+      try {
+        ObjectOutputStream(outputStream).use { objectOutputStream ->
+          // Send total number of files which will be transferred
+          objectOutputStream.writeObject("" + wifiDirectManager.totalFilesForTransfer)
+          val fileItemArrayList =
+            wifiDirectManager.getFilesForTransfer()
+          for (fileItem in fileItemArrayList) { // Send the names of each of those files, in order
+            objectOutputStream.writeObject(fileItem.fileName)
+            Log.d(TAG, "Sending " + fileItem.fileUri.toString())
+          }
+        }
+      } catch (e: Exception) {
+        e.printStackTrace()
+      }
+    } else { // Device is not the file sender
+      try {
+        ObjectInputStream(inputStream).use { objectInputStream ->
+          // Read the number of files
+          val totalFilesObject = objectInputStream.readObject().toString()
+          if (totalFilesObject.javaClass == String::class.java) {
+            val total: Int = totalFilesObject.toInt()
+            if (BuildConfig.DEBUG) Log.d(TAG, "Metadata: $total files")
+            val fileItems = ArrayList<FileItem>()
+            for (i in 0 until total) { // Read names of each of those files, in order
```

Comment:
we don't really care about i, we can use `repeat(n) { }` to keep doing an action n times

Reply:
actually the output is a list, we should just map the range of the for loop

Reply:
@gouri-panda just drawing your attention here incase you missed my 2nd comment

Reply:
@macgills I don't know how to use map in this case?

Reply:
done